src/main/java/net/minecraft/server/Block.java

```diff
@@ -301,23 +302,35 @@ public final void g(World world, int i, int j, int k, int l) {
         this.dropNaturally(world, i, j, k, l, 1.0F);
     }
 
-    public void dropNaturally(World world, int i, int j, int k, int l, float f) {
+    public final void dropNaturally(World world, int i, int j, int k, int l, float f) {
```

Comment:
I can't think of a reason that a plugin would want to create a class that extends NMS block, as the only use I can think of for it would be to add a new block which wouldn't work with vanilla clients. If you think it should be changed back that's fine, it won't break anything to have that method not be final.

Reply:
Spout is one of the plugins that extends some core NMS classes, such as NetServerHandler. I think it could extend Block too, I'm not too sure.

Reply:
100% buck wild if it does this. I've changed it back just in case. I mostly changed it so my IDE would rat out other NSM blocks that used dropNaturally to do weird drops like crops and beds so I could change them to my new system.

Reply:
@Moresteck @moderatorman Likely out of the scope of this issue

I'm very conflicted as to what extent we should aim at not refactoring NMS classes due to plugins like Citizens and Spout and especially noting that methods such as setAccessible utilized by reflections are deprecated in Java 9.

Refactoring the classes makes a more readable code base and could simplify future development so we will have to determine the risk versus reward I guess.

Should servers be responsible for maintaining plugins such as Spout and Citizen if NMS changes occur in Poseidon?

I imagine Citizens is used a bit, however, is Spout used much within the community? I know the main thing it was used for in the past was packet events but Poseidon already has a proper packet event system. Knowing this, should we be worried about that?

Thoughts?

Reply:
I think it really depends on what we want to do with Project Poseidon.

1. Running contemporary plugins that may depend on old NMS code
This gives us an existing pool of plugins to run at the expense of Windows levels of back-compat if we don't want them to break.

2. Facilitating the development of new plugins
Idealy, these new plugins wouldn't touch NMS at all. If a dev requires a feature that isn't avaliable though the API, they can add the feature to Poseidon and open a pr.
If a dev does decide to use NMS, the do so with the understanding that the plugin may break in future releases of the server and need to be updated over time.
This means we have a limited pool of existing plugins to use but we don't need to worry as much about keeping compatability with old NMS code or even old APIs.

I'm partial to option 2, though I understand that a lot of Poseidon users are probably trying to recreate the experience of playing Minecraft beta 1.7.3 multiplayer including the old plugins they would have been using at the time.

Reply:
> @Moresteck @moderatorman Likely out of the scope of this issue
> 
> I'm very conflicted as to what extent we should aim at not refactoring NMS classes due to plugins like Citizens and Spout and especially noting that methods such as setAccessible utilized by reflections are deprecated in Java 9.
> 
> Refactoring the classes makes a more readable code base and could simplify future development so we will have to determine the risk versus reward I guess.
> 
> Should servers be responsible for maintaining plugins such as Spout and Citizen if NMS changes occur in Poseidon?
> 
> I imagine Citizens is used a bit, however, is Spout used much within the community? I know the main thing it was used for in the past was packet events but Poseidon already has a proper packet event system. Knowing this, should we be worried about that?
> 
> Thoughts?

I think we ought to leave it be, frankly. The code really isn't all that hard to understand. I'm assuming everyone here is a skilled programmer, and won't be needing such a dramatic change just for convenience. Backwards compatibility is far more important than readability in the grand scheme of things, especially when it comes to things like Spout.

I realize replacements can be made, but let's not forget that people still use vanilla CB1060 in some cases, and if they ever switch over, the transition ought to be as seamless as possible. That was one of the primary reasons my server started using it, and I feel most people here might have had a difference of opinion if they had to change all of their plugins along with their server software. Accessibility and backwards compatibility should take priority. That's my two cents.

Reply:
@Meep3692, I'm in favor of incorporating this feature, but could you please retain the original method and variable names if feasible?

Regarding the refactoring debate, I acknowledge the points raised by Code and Moderator, yet I'm uncertain about imposing such restrictions on Poseidon, especially considering that I doubt many servers employ plugins that depend on NMS or reflection (which, ideally, shouldn't be used anyway).

I'd love to revisit this topic later this year when we have more time, outside of this current PR discussion.

Reply:
I have reverted the last of the name changes. The only changes in NMS now are the addition of getDrops and dropNaturally's use of it. Some of the Block subclasses look a bit strange in the diff as they no longer override dropNaturally (should look the same to reflection though) in favour of the new system that makes these changes possible.